cfgrammar/src/lib/yacc/parser.rs

```diff
@@ -362,117 +360,78 @@ fn add_duplicate_occurrence(
 
 /// The actual parser is intended to be entirely opaque from outside users.
 impl YaccParser<'_> {
-    pub(crate) fn new(header: &mut Header, src: String) -> YaccParser {
+    pub(crate) fn new(yacc_kind: YaccKind, src: &str) -> YaccParser {
         YaccParser {
-            header,
-            yacc_kind: None,
+            yacc_kind,
             src,
             num_newlines: 0,
             ast: GrammarAST::new(),
             global_actiontype: None,
         }
     }
 
+    #[cfg(test)]
+    pub(crate) fn from_src(src: &str) -> Result<YaccParser, Vec<HeaderError>> {
```

Comment:
This one we mostly needed for the testsuite, because we had lost any internal methods that lack a `YaccKind`.
Perhaps these tests should be moved elsewhere.

Reply:
It sounds like it's worth trying. If that move simplifies the code a bit, we probably have a winner!